lib/elixir/lib/application.ex

```diff
@@ -734,9 +738,17 @@ defmodule Application do
   """
   @spec delete_env(app, key, timeout: timeout, persistent: boolean) :: :ok
   def delete_env(app, key, opts \\ []) when is_atom(app) do
+    maybe_warn_on_app_env_key(key)
     :application.unset_env(app, key, opts)
   end
 
+  # TODO: turn warning into error on Elixir v2.0
+  defp maybe_warn_on_app_env_key(key) when is_atom(key),
+    do: :ok
+
+  defp maybe_warn_on_app_env_key(key),
+    do: IO.warn("passing non-atom as application env key is deprecated (got: `#{inspect(key)}`)")
```

Comment:
Those are runtime warnings, so we probably want to emit them only once per key. See the private function `IO.warn_once` and how we use it in the codebase for more info.